File: EticaretJava/app/src/main/java/com/example/eticaretjava/repo/Factories.java
package com.example.eticaretjava.repo;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.example.eticaretjava.data.TokenStore;
import com.example.eticaretjava.viewmodel.AuthViewModel;
import com.example.eticaretjava.viewmodel.CartViewModel;
import com.example.eticaretjava.viewmodel.HomeViewModel;
import com.example.eticaretjava.viewmodel.OrdersViewModel;
import com.example.eticaretjava.viewmodel.ProductDetailViewModel;

public class Factories {

    public static class AuthVMFactory implements ViewModelProvider.Factory {
        private final TokenStore tokenStore;
        private final RepositoriesImpl.AuthRepositoryImpl repo;

        public AuthVMFactory(Context ctx) {
            tokenStore = new TokenStore(ctx.getApplicationContext());
            repo = new RepositoriesImpl.AuthRepositoryImpl(RetrofitProvider.authApi, tokenStore);
        }

        @NonNull
        @Override
        @SuppressWarnings("unchecked")
        public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
            return (T) new AuthViewModel(repo);
        }
    }

    public static class HomeVMFactory implements ViewModelProvider.Factory {
        private final RepositoriesImpl.ProductRepositoryImpl repo =
                new RepositoriesImpl.ProductRepositoryImpl(RetrofitProvider.productApi);

        @NonNull
        @Override
        @SuppressWarnings("unchecked")
        public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
            return (T) new HomeViewModel(repo);
        }
    }

    public static class ProductDetailVMFactory implements ViewModelProvider.Factory {
        private final RepositoriesImpl.ProductRepositoryImpl prodRepo =
                new RepositoriesImpl.ProductRepositoryImpl(RetrofitProvider.productApi);

        private final RepositoriesImpl.CartRepositoryImpl cartRepo;

        public ProductDetailVMFactory(Context ctx) {
            cartRepo = new RepositoriesImpl.CartRepositoryImpl(
                    RetrofitProvider.cartApi,
                    new TokenStore(ctx.getApplicationContext())
            );
        }

        @NonNull
        @Override
        @SuppressWarnings("unchecked")
        public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
            return (T) new ProductDetailViewModel(prodRepo, cartRepo);
        }
    }

    public static class CartVMFactory implements ViewModelProvider.Factory {
        private final RepositoriesImpl.CartRepositoryImpl repo;

        public CartVMFactory(Context ctx) {
            repo = new RepositoriesImpl.CartRepositoryImpl(
                    RetrofitProvider.cartApi,
                    new TokenStore(ctx.getApplicationContext())
            );
        }

        @NonNull
        @Override
        @SuppressWarnings("unchecked")
        public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
            return (T) new CartViewModel(repo);
        }
    }

    public static class OrdersVMFactory implements ViewModelProvider.Factory {
        private final RepositoriesImpl.OrderRepositoryImpl repo;

        public OrdersVMFactory(Context ctx) {
            repo = new RepositoriesImpl.OrderRepositoryImpl(
                    RetrofitProvider.orderApi,
                    new TokenStore(ctx.getApplicationContext())
            );
        }

        @NonNull
        @Override
        @SuppressWarnings("unchecked")
        public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
            return (T) new OrdersViewModel(repo);
        }
    }
}

File: EticaretKotlin/app/src/main/java/com/example/eticaretkotlin/repo/Factories.kt
package com.example.eticaretkotlin.repo

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.example.eticaretkotlin.data.TokenStore

import com.example.eticaretkotlin.viewmodel.AuthViewModel
import com.example.eticaretkotlin.viewmodel.CartViewModel
import com.example.eticaretkotlin.viewmodel.HomeViewModel
import com.example.eticaretkotlin.viewmodel.OrdersViewModel
import com.example.eticaretkotlin.viewmodel.ProductDetailViewModel


class AuthVMFactory(ctx: Context) : ViewModelProvider.Factory {
    private val tokenStore = TokenStore(ctx.applicationContext)

    // ✅ interface tipinde tut
    private val repo: AuthRepositoryImpl =
        AuthRepositoryImpl(RetrofitProvider.authApi, tokenStore)

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        @Suppress("UNCHECKED_CAST")
        return AuthViewModel(repo) as T
    }
}


class HomeVMFactory : ViewModelProvider.Factory {
    private val repo = ProductRepositoryImpl(RetrofitProvider.productApi) // kendi impl’in
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        @Suppress("UNCHECKED_CAST") return HomeViewModel(repo) as T
    }
}

class ProductDetailVMFactory(ctx: Context) : ViewModelProvider.Factory {
    private val prod = ProductRepositoryImpl(RetrofitProvider.productApi)
    private val cart = CartRepositoryImpl(RetrofitProvider.cartApi, tokenStore = TokenStore(ctx.applicationContext))
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        @Suppress("UNCHECKED_CAST") return ProductDetailViewModel(prod, cart) as T
    }
}


class CartVMFactory(ctx: Context) : ViewModelProvider.Factory {
    private val repo: CartRepository =
        CartRepositoryImpl(RetrofitProvider.cartApi, TokenStore(ctx.applicationContext))

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        @Suppress("UNCHECKED_CAST")
        return CartViewModel(repo) as T
    }
}


class OrdersVMFactory(ctx: Context) : ViewModelProvider.Factory {
    private val repo = OrderRepositoryImpl(RetrofitProvider.orderApi, TokenStore(ctx.applicationContext))
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        @Suppress("UNCHECKED_CAST") return OrdersViewModel(repo) as T
    }
}


